src/app/use-cases/status-signature-request.ts

```diff
@@ -0,0 +1,123 @@
+import * as E from "fp-ts/Either";
+import * as t from "io-ts";
+
+import { pipe } from "fp-ts/function";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+
+import { ProblemDetail } from "@pagopa/handler-kit/lib/problem-detail";
+import { SignatureRequest } from "../../signature-request/signature-request";
+
+export class ActionNotAllowedError extends Error implements ProblemDetail {
+  type = "/problems/action-not-allowed";
+  title = "Action not allowed";
+  status = "400";
+  constructor(public detail: string) {
+    super(detail);
+    this.name = "ActionNotAllowedError";
+  }
+}
+
+export const SignatureRequestAction = t.keyof({
+  UPLOAD_DOCUMENT: null,
+  MARK_AS_READY: null,
+  VALIDATE_DOCUMENT: null,
+  MARK_AS_SIGNED: null,
+});
+export type SignatureRequestAction = t.TypeOf<typeof SignatureRequestAction>;
+
+export const nextStatus =
+  (request: SignatureRequest) =>
+  (
+    action: SignatureRequestAction
+  ): E.Either<ActionNotAllowedError, SignatureRequest> => {
+    switch (request.status) {
+      case "DRAFT":
+        return pipe(action, whenDraft(request));
+      case "WAIT_FOR_ISSUER":
+        return pipe(action, whenWaitForIssuer(request));
+      case "READY":
+        return pipe(action, whenReady(request));
+      case "WAIT_FOR_SIGNATURE":
+        return pipe(action, whenWaitForSignature(request));
+      case "SIGNED":
+        return E.left(
+          new ActionNotAllowedError(
+            "This action is not allowed when signature request is in already SIGNED!"
+          )
+        );
+      default:
+        return E.left(new ActionNotAllowedError("Status not valid!"));
+    }
+  };
+
+const whenDraft =
+  (request: SignatureRequest) =>
+  (
+    action: SignatureRequestAction
+  ): E.Either<ActionNotAllowedError, SignatureRequest> =>
+    action === "UPLOAD_DOCUMENT"
```

Comment:
I'd use a switch here as well

Reply:
It reminds to me this https://xstate.js.org/docs/ (or redux). Maybe we can consider to use xstate for this kind of state machine (but you have to create a singleton state machine then pass it to the methods that need it)

Reply:
Another use of xstate is logging, see `onTransition` here: https://xstate.js.org/docs/

Reply:
@lucacavallaro what do you think about it?

Reply:
lint suggests: `switch statements should have at least 3 case clauseses` for this reason I used an inline check.

...
For this type of implementation I took a cue from xstate, but it seemed excessive to use the entire library.

Reply:
I agree more with sonarcloud:
<img width="733" alt="image" src="https://user-images.githubusercontent.com/148224/190187447-70ebfc9e-80b6-4ce0-bd8a-af30ae0ee58e.png">

Reply:
ok, I take the control off!